plugins/Phaser/PhaserEffect.cpp

```diff
@@ -0,0 +1,623 @@
+				{
+					m_currentPeak[i] = qMax(m_currentPeak[i] * m_relCoeff, sAbs[i]);
+				}
+
+				m_currentPeak[i] = qBound(PHA_NOISE_FLOOR, m_currentPeak[i], 10.0f);
+			}
+		}
+		else
+		{
+			m_currentPeak[0] = 0;
+			m_currentPeak[1] = 0;
+		}
+
+		// Calculate real allpass filter frequencies
+		float lfoResults[2] = {0, 0};
+		if (enableLFO)
+		{
+			m_lfo->tick(&lfoResults[0], &lfoResults[1]);
+		}
+
+		const float realLfo[2] = {lfoResults[0] * amount, lfoResults[1] * amount};
+		const float realInFollow[2] = {m_currentPeak[0] * inFollow, m_currentPeak[1] * inFollow};
+		for (int i = 0; i < 2; i++)
+		{
+			m_displayCutoff[i] = qBound(20.f, detuneWithOctaves(cutoff, realLfo[i] + realInFollow[i]), 20000.f);
+			m_realCutoff[i] = qBound(20.f, detuneWithOctaves(cutoff, (realLfo[i] + realInFollow[i]) * cutoffControl), 20000.f);
+		}
+		
+		/*
+		Precalculate the allpass filter coefficients.
+		Because of very aggressive CPU optimization, some of
+		the coefficient names are inaccurate concerning what
+		they actually do.
+		*/
+		float b0[2];
+		float b1[2];
+		for (int i = 0; i < 2; ++i)
+		{
+			const float w0 = m_twoPiOverSr * m_realCutoff[i];
+			const float a0 = 1 + (sin(w0) / (resonance * 2.f));
+			b0[i] = (1 - (a0 - 1)) / a0;
+			b1[i] = (-2*cos(w0)) / a0;
+		}
+
+		float firstAdd[2] = {0, 0};
+		float secondAdd[2] = {0, 0};
+		float delayOut[2] = {0, 0};
+
+		for (int i = 0; i < 2; ++i)
+		{
+			// Calculate delay amount to find read location
+			float readLoc = m_filtFeedbackLoc -
+				qMin(abs(delay + ((realLfo[i] + realInFollow[i]) * delayControl)), 20.f) *
+				0.001f * Engine::mixer()->processingSampleRate();
+
+			if (readLoc < 0) {readLoc += m_delayBufSize;}
+			float readLocFrac = fraction(readLoc);
+
+			// Read value from delay buffer
+			if (readLoc < m_delayBufSize - 1)
+			{
+				delayOut[i] = m_filtDelayBuf[i][floor(readLoc)] * (1 - readLocFrac) + m_filtDelayBuf[i][ceil(readLoc)] * readLocFrac;
+			}
+			else// For when the interpolation wraps around to the beginning of the buffer
+			{
+				delayOut[i] = m_filtDelayBuf[i][m_delayBufSize - 1] * (1 - readLocFrac) + m_filtDelayBuf[i][0] * readLocFrac;
+			}
+		}
+
+		float tempAdd[2] = {0, 0};
+		switch (mode)
+		{
+			case 0: case 1:
+			{
+				tempAdd[0] = delayOut[0] * feedback;
+				tempAdd[1] = delayOut[1] * feedback;
+				break;
+			}
+			case 2:
+			{
+				tempAdd[0] = lastSecondAdd[0] * feedback;
+				tempAdd[1] = lastSecondAdd[1] * feedback;
+				break;
+			}
+			case 3:
+			{
+				tempAdd[0] = delayOut[0] * feedback;
+				tempAdd[1] = delayOut[1] * feedback;
+				for (int i = 0; i < order; ++i)
+				{
+					for (int j = 0; j < 2; ++j)
+					{
+						for (int k = 0; k < 1 + doubleVal; ++k)// Runs twice if doubleVal is true
+						{
+							tempAdd[j] = calcAllpassFilter(tempAdd[j], sample_rate, i, j, b0[j], b1[j]);
+						}
+					}
+				}
+			}
+		}
+		if (analog)
+		{
+			firstAdd[0] = tanh(s[0] * analogDist) / analogDist + tanh(tempAdd[0] * analogDist) / analogDist;
```

Comment:
It isn't called "analog" because the distortion is analog, it's called "analog" because having distortion inside of the feedback loop is a characteristic of analog audio hardware.  Like how Moog filters work by adding nonlinearities (distortion) into the filter feedback loops.

That being said, I do agree that aliasing should be avoided... but I am of the opinion that oversampling should be handled by LMMS itself rather than individual plugins, and that's out-of-scope for this plugin.  I think every plugin should come with oversampling options, just like how every plugin comes with a Wet/Dry knob.

So, unless you disagree, I don't think we should bother with putting oversampling into here.  I didn't put it into any of my other plugins for the same reason.  If we implement oversampling as an LMMS feature rather than a per-plugin feature, we'll never have to worry about it ever again... because, in reality, over 90% of the plugins in the entire catalog could benefit from oversampling, so adding it in on a per-plugin basis seems quite unnecessary.

Reply:
Aliasing is a very important fundamental concept in DSP, and has many reasons to be avoided, harmonic sounds become inharmonic, sound out of tune when played with other notes, fill the audio spectrum with unwanted content that makes final mixdowns muddy, and are an indicator of the quality of the product. I am aware that historically aliasing has not been of a high priority in LMMS, but I am not convinced adding to the problem would be constructive.

It is good practice to only oversample where necessary, due to the increased processing, If LMMS were to have a user option to oversample by plugin, this would be very inefficient due to the whole signal path in the plugin being processed multiple times, and expecting the artist to have an understanding of DSP and the consequences of setting the oversample rate. This is why I am not aware of any reputable DAW that takes this approach. It is always up to the plugin developer to handle anti-aliasing in an efficient manner.

That leaves the issue of how to progress with this plugin, specifically the "Analog distortion", I have a few suggestions, any one of these alone would solve this issue.

1. Oversample and decimate the non linear processing (tanh). 
2. Rename the option from "Analog" to better describe the outcome, I would suggest a name that makes it clear that the aliasing is intended, as you have done elsewhere.
3. Leave it as is, and hope no one notices or cares.

I shall leave the decision up to yourself, and shall not raise this issue again on this PR, unless asked.

Reply:
I agree that we should try to reduce aliasing, but there are some issues.
> Oversample and decimate the non linear processing (tanh).

AFAIK, that add some delays around the non-linear processing. Wouldn't that affect the outcome in a way other than reducing aliasing?
Slightly out of scope, but I think filters may also leverage from oversampling for better frequency/phase response.

> Rename the option from "Analog" to better describe the outcome

That might be a good choice, but I have no ideas on better names.